pandas/core/indexes/base.py

```diff
@@ -5681,9 +5681,7 @@ def _should_fallback_to_positional(self) -> bool:
         """
         Should an integer key be treated as positional?
         """
-        if isinstance(self.dtype, np.dtype) and self.dtype.kind in ["i", "u", "f"]:
-            return False
-        return not self._holds_integer()
+        return not is_numeric_dtype(self.dtype) or is_bool_dtype(self.dtype)
```

Comment:
does this impact perf?  the is_foo_dtype checks are pretty slow

Reply:
It probably does a bit. For `float64` I get 1.71 µs, would probably  be faster to use `infer_dtype` (after #51054).

I cabn do a timing after #51054 has been merged.

Reply:
I looked into using inferred and it is better, 120 ns now, vs 1000 ns in first version.